R/filter_state_api_NEW.R

```diff
@@ -0,0 +1,216 @@
+#' filter_two <- filter_var("dataname1", "varname2", 1:10, 2, "integers", TRUE, FALSE, extra2 = "extratwo")
+#' filter_three <- filter_var("dataname2", "varname3", 1:10/10, 0.2, "doubles", TRUE, FALSE, extra1 = "extraone", extra2 = "extratwo")
+#'
+#' @export
+#' @rdname new_api
+#'
+filter_var <- function(
+    dataname,
+    varname,
+    choices = NULL,
+    selected = NULL,
+    varlabel = character(0),
+    keep_na = NULL,
+    keep_inf = NULL,
+    fixed = FALSE,
+    ...
+) {
+  checkmate::assert_string(dataname)
+  checkmate::assert_string(varname)
+  checkmate::assert_atomic(choices)
+  checkmate::assert_atomic(selected)
+  if (!is.null(choices) && !is.null(selected)) {
+    if (inherits(choices, c("integer", "numeric", "Date", "POSIXt"))) {
+      rc <- range(choices, na.rm = TRUE)
+      rs <- range(selected, na.rm = TRUE)
+      checkmate::assert_true(
+        rs[1L] >= rc[1L] && rs[2L] <= rc[2L],
+        .var.name = "range of \"selected\" is within range of \"choices\""
+      )
+    } else if (inherits(choices, c("logical", "character", "factor"))) {
+      checkmate::assert_subset(selected, choices)
+    } else {
+      stop("filter_var cannot handle \"choices\" of type: ", toString(class(choices)))
+    }
+  }
+  checkmate::assert_character(varlabel, max.len = 1L)
+  checkmate::assert_flag(keep_na, null.ok = TRUE)
+  checkmate::assert_flag(keep_inf, null.ok = TRUE)
+  checkmate::assert_flag(fixed)
+
+  ans <- list(
+    dataname = dataname,
+    varname = varname,
+    choices = choices,
+    selected = selected,
+    varlabel = varlabel,
+    keep_na = keep_na,
+    keep_inf = keep_inf,
+    fixed = fixed
+  )
+  extras <- list(...)
+  if (length(extras) != 0L) ans$extras <- extras
+
+  class(ans) <- c("teal_slice", class(ans))
+  ans
+}
+
+c.teal_slice <- function(...) {
+  ans <- unlist(list(...), recursive = FALSE)
+  class(ans) <- "teal_slice"
+  ans
+}
+print.teal_slice <- function(x) {
+  str(x)
+}
+
+
+
+#' complete filter settings
+#'
+#' Collate single variable filter states into a complete filter specification.
+#'
+#' @param ... any number of `teal_slice` objects
+#' @param slices optional `list` of `teal_slice` objects
+#' @exclude `named list` of `character` vectors where list names match names of data sets
+#'          and vectors elements match variable names in respective data sets;
+#'          specifies which variables are not allowed to be filtered
+#' @count_type `character(1)` string specifying how observations are tallied by these filter states
+#'
+#' @return
+#' Object of class `teal_slices`, which is an unnamed list of `teal_slice` objects.
+#'
+#' @examples
+#' all_filters <- filter_settings(
+#'   filter_one,
+#'   filter_two,
+#'   filter_three,
+#'   exclude = list(
+#'     "dataname1" = "varname2"
+#'   )
+#' )
+#'
+#' @export
+#' @rdname new_api
+#'
+filter_settings <- function(
+    ...,
+    slices,
```

Comment:
I see the point but I think we can make it simpler. I don't like the idea to give user more opportunities
```suggestion
```

Reply:
I have a proposition for you
```
arg_list <- slices
arg_list <- c(arg_list, list(exclude = excluded_varnames, count_type = private$count_type)

do.call(filter_settings, arg_list)
```

Reply:
:+1: